### DISCUSSION TITLE
Able to test http call

### DISCUSSION THREAD

**Discussion by KenBonny** [fsprojects/FsHttp#119] [Ideas]
I use [Flurl](https://flurl.dev/) extensively when I use C#. I don't like that there is not an immutable builder, but I do like that I can easily [test it](https://flurl.dev/docs/testable-http/). Would it be possible to have something similar for FsHttp?

**Comment by SchlenkR** [comment]
In general, I prefer provide functionality that enables to build a testing library on top of FsHttp rather than providing something from inside.

Currently, it is possible to register transformation- or factory functions for HttpRequestMessage and HttpClient (either globally or per-request). The step that actually sends the request (i.e. transforms the request into a response) is not interceptable (might be a feature we could think about).

Using the interception possibilities for HttpRequestMessage / HttpClient, it might be possible to prototype something that you look for.

**Reply by SchlenkR** [reply]
I have something here:

https://github.com/fsprojects/FsHttp/blob/d377a1f3c3eb239dc69eab72cfa438af34e0bc04/fiddle/discuss-119-Testing.fsx

It is just a quick prototype, but shows you how you can use Suave (a super slick F# webserver) and FsHttp to rewrite requests like this:

```fsharp
    // Intercept a GET on wikipedia and return 200 with content "You know it"
    use _ =
        [
            {
                method = Filters.GET
                url = Uri("https://www.wikipedia.de")
                rewrite = "/"
                resp = OK "You know it" 
            }
        ]
        |> rewire

    let responseText = 
        http {
            GET "https://www.wikipedia.de"
        }
        |> Request.send
        |> Response.toText
    
    if responseText <> "You know it" then
        failwith "Test failed."
```

Please be aware that this solution is not thread safe in any way and just a demo :)

**Reply by KenBonny** [reply]
This is really cool and I was not aware I could do this. Thanks for showing me.

**Reply by SchlenkR** [reply]
You’re welcome!
